File: wrappers/vb/LibRangeMap/BooleanRangeMapper.vb
Imports System.Text.Json
Imports System.Text.Json.Serialization

Public Class BooleanRangeMapper
    Private Shared ReadOnly JsonOptions As New JsonSerializerOptions With {
        .PropertyNamingPolicy = Nothing,
        .WriteIndented = False
    }

    Private ReadOnly _outputMin As Double
    Private ReadOnly _outputMax As Double
    Private ReadOnly _falseValue As Double
    Private ReadOnly _trueValue As Double

    Public Sub New(Optional outputMin As Double = -1.0, Optional outputMax As Double = 1.0, Optional falseValue As Double = -1.0, Optional trueValue As Double = 1.0)
        If Double.IsNaN(outputMin) OrElse Double.IsInfinity(outputMin) OrElse Double.IsNaN(outputMax) OrElse Double.IsInfinity(outputMax) OrElse Double.IsNaN(falseValue) OrElse Double.IsInfinity(falseValue) OrElse Double.IsNaN(trueValue) OrElse Double.IsInfinity(trueValue) Then
            Throw New ArgumentException("all range values must be finite.")
        End If
        If outputMin >= outputMax Then
            Throw New ArgumentException("outputMin must be less than outputMax.")
        End If
        If falseValue < outputMin OrElse falseValue > outputMax OrElse trueValue < outputMin OrElse trueValue > outputMax Then
            Throw New ArgumentException("boolean values must lie within the output range.")
        End If
        If falseValue = trueValue Then
            Throw New ArgumentException("falseValue and trueValue must differ.")
        End If

        _outputMin = outputMin
        _outputMax = outputMax
        _falseValue = falseValue
        _trueValue = trueValue
    End Sub

    Public Function MapValue(value As Boolean) As Double
        Return If(value, _trueValue, _falseValue)
    End Function

    Public Function Map(value As Boolean) As Double
        Return MapValue(value)
    End Function

    Public Function Spec() As BooleanMapperSpec
        Return New BooleanMapperSpec With {
            .SpecVersion = CurrentSpecVersion,
            .MapperType = "boolean_range",
            .OutputRange = New Double() {_outputMin, _outputMax},
            .FalseValue = _falseValue,
            .TrueValue = _trueValue
        }
    End Function

    Public Function ToJson() As String
        Return JsonSerializer.Serialize(Spec(), JsonOptions)
    End Function

    Public Shared Function FromJson(json As String) As BooleanRangeMapper
        Dim spec = JsonSerializer.Deserialize(Of BooleanMapperSpec)(json, JsonOptions)
        If spec Is Nothing Then
            Throw New InvalidOperationException("mapper spec JSON produced null.")
        End If
        Return FromSpec(spec)
    End Function

    Public Shared Function FromSpec(spec As BooleanMapperSpec) As BooleanRangeMapper
        If spec.SpecVersion <> CurrentSpecVersion Then
            Throw New ArgumentException($"unsupported spec_version '{spec.SpecVersion}'.", NameOf(spec))
        End If
        If spec.MapperType <> "boolean_range" Then
            Throw New ArgumentException($"unsupported mapper_type '{spec.MapperType}'.", NameOf(spec))
        End If
        If spec.OutputRange Is Nothing OrElse spec.OutputRange.Length <> 2 Then
            Throw New ArgumentException("output_range must contain exactly two values.", NameOf(spec))
        End If
        Return New BooleanRangeMapper(spec.OutputRange(0), spec.OutputRange(1), spec.FalseValue, spec.TrueValue)
    End Function
End Class

File: wrappers/csharp/LibRangeMap/BooleanRangeMapper.cs
using System;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace LibRangeMap;

public sealed class BooleanRangeMapper
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = null,
        WriteIndented = false
    };

    private const string SpecVersion = MapperSpec.CurrentSpecVersion;
    private const string SpecType = "boolean_range";

    public double OutputMin { get; }
    public double OutputMax { get; }
    public double FalseValue { get; }
    public double TrueValue { get; }
    public string? Name { get; }

    public BooleanRangeMapper(
        double outputMin = -1.0,
        double outputMax = 1.0,
        double? falseValue = null,
        double? trueValue = null,
        string? name = null)
    {
        ValidateFinite(nameof(outputMin), outputMin);
        ValidateFinite(nameof(outputMax), outputMax);

        if (outputMax <= outputMin)
        {
            throw new ArgumentException("output_range must be ordered from lower value to higher value.");
        }
        if (name is not null && name.Length == 0)
        {
            throw new ArgumentException("name must not be empty.", nameof(name));
        }

        OutputMin = outputMin;
        OutputMax = outputMax;
        FalseValue = falseValue ?? outputMin;
        TrueValue = trueValue ?? outputMax;
        Name = name;
    }

    public double MapValue(bool value) => value ? TrueValue : FalseValue;

    public double Map(bool value) => MapValue(value);

    public string ToJson() => JsonSerializer.Serialize(ToSpec(), JsonOptions);

    private BooleanMapperSpec ToSpec() => new()
    {
        SpecVersion = SpecVersion,
        MapperType = SpecType,
        OutputRange = new[] { OutputMin, OutputMax },
        FalseValue = FalseValue,
        TrueValue = TrueValue,
        Name = Name,
    };

    public static BooleanRangeMapper FromJson(string json)
    {
        var spec = JsonSerializer.Deserialize<BooleanMapperSpec>(json, JsonOptions)
            ?? throw new InvalidOperationException("mapper spec JSON produced null.");
        return FromSpec(spec);
    }

    public static BooleanRangeMapper FromSpec(BooleanMapperSpec spec)
    {
        if (spec.SpecVersion != SpecVersion)
        {
            throw new ArgumentException($"unsupported spec_version '{spec.SpecVersion}'.", nameof(spec));
        }
        if (spec.MapperType != SpecType)
        {
            throw new ArgumentException($"unsupported mapper_type '{spec.MapperType}'.", nameof(spec));
        }
        if (spec.OutputRange is null || spec.OutputRange.Length != 2)
        {
            throw new ArgumentException("output_range must contain exactly two values.", nameof(spec));
        }
        if (double.IsNaN(spec.FalseValue) || double.IsInfinity(spec.FalseValue)
            || double.IsNaN(spec.TrueValue) || double.IsInfinity(spec.TrueValue))
        {
            throw new ArgumentException("false_value and true_value must be finite.");
        }

        return new BooleanRangeMapper(
            spec.OutputRange[0],
            spec.OutputRange[1],
            spec.FalseValue,
            spec.TrueValue,
            spec.Name);
    }

    private static void ValidateFinite(string label, double value)
    {
        if (double.IsNaN(value) || double.IsInfinity(value))
        {
            throw new ArgumentException($"{label} must be a finite number.");
        }
    }
}

public sealed class BooleanMapperSpec
{
    [JsonPropertyName("spec_version")]
    public string SpecVersion { get; init; } = MapperSpec.CurrentSpecVersion;

    [JsonPropertyName("mapper_type")]
    public string MapperType { get; init; } = "boolean_range";

    [JsonPropertyName("output_range")]
    public double[] OutputRange { get; init; } = new double[2];

    [JsonPropertyName("false_value")]
    public double FalseValue { get; init; }

    [JsonPropertyName("true_value")]
    public double TrueValue { get; init; }

    [JsonPropertyName("name")]
    public string? Name { get; init; }
}
